File: plot_param_panel.m
%function plot_result(DIwave,t,p,simMEP,ref,spike_times,NRMSD,Rm,Hm,Hrc)
function plot_param_panel(p,ref)

    spike_times=ref.sim.spike_times;
    NRMSD=ref.NRMSD;
    R=ref.model.R;
    Wexc=ref.model.Wexc;
    RWinh=ref.model.RWinh; % R*Winh
    simMEP=ref.sim.simMEP2;
    DIwave=ref.model.DIwave;
    t=ref.model.t;

    nParams=length(p);
    figure;
    if isfield(ref.model,'boundary')
      boundary=ref.model.boundary;
    end
     % boundary=[zeros(nParams,1),max(1,p(:)*1.1)] ; 


    ttext=ref.model.boundarytext;
    for i=1:nParams
       subplot(3,nParams,i);hold on
       rescale=0.9;
       if isfield(ref.model,'boundary')
          locP=rescale*(p(i)-boundary(i,1))/(boundary(i,2)-boundary(i,1));% map p into 0~1
       else
          locP=0.5*rescale;
       end
       scatter(1,locP,"filled")
       text(1.3,locP,sprintf('%.2f',p(i)))
       if isfield(ref.model,'boundary')
           scatter([1,1],[0,rescale],"k_")
           plot([1,1],[0,rescale],'k')
           text(1,0,num2str(boundary(i,1)),'HorizontalAlignment','center','VerticalAlignment','top')
           text(1,rescale,num2str(boundary(i,2)),'HorizontalAlignment','center','VerticalAlignment','bottom')     
       end
       ylim([-0.1 1.1])
       axis off
       title(ttext{min(i,length(ttext))})
    end

    subplot(312);
    plot(ref.y0(:),'k','linewidth',2);hold on;
    plot(simMEP(:),'r','linewidth',1);
    R2=1-sumsqr(ref.y0(:)-simMEP(:))/sumsqr(ref.y0(:)-mean(ref.y0(:)));
    fprintf('R^2 = %g\n',R2)
    title(sprintf('MEPs of subject %d \n(R^2 = %.2g, NRMSD = %.2g%%)',ref.subj,R2,NRMSD*100))

    plot([1;1]*(0:size(simMEP,2)-1)*size(simMEP,1),ylim,'k')
    xlim([0 length(simMEP(:))])
    xlabel(sprintf('Time (%g-%g ms)',ref.tcrop))
    set(gca,'xtick',(1:size(simMEP,2)-1)*size(simMEP,1),'xticklabel',[]);
    ylimit=ylim;
    for i=1:length(ref.intensities)
        text(length(ref.t0)*(i-0.5),ylimit(2)*0.8,sprintf('%d%%\nMSO',ref.intensities(i)),'horizontalalignment','center','fontsize',8)
    end
    lgd=legend('MEP','simMEP','location','west');
    lgd.ItemTokenSize = [5,1];
    ylabel('Amplitude (mV)');
    
    
    subplot(3,4,9)
    sel=[length(ref.intensities),round(length(ref.intensities)/2),1]; % select 3 TMS intensities to show
    legend_labels = strcat(string(ref.intensities(sel)), '%');
    plot(t,DIwave(sel,:)','linewidth',1.5)
    xlabel('Time (ms)')
    title('DIwaves')
    xlim([0 20]);
    ylimit=ylim; ylim([-0.05 ylimit(2)]);
    lgd=legend(legend_labels,'location','northeast');%legend(num2str(ref.intensities(sel)'),'location','northeast')
    lgd.ItemTokenSize = [5,1];

    subplot(3,4,10)
    h1=plot(1:100,R,'k.');title('Model param.');xlabel('Motor neuron')
    hold on;plot([1 10 20 60 100],R([1 10 20 60 100]),'ko'); 
    
    if ref.model.withRC
        h2=plot(1:100,Wexc,'b','linewidth',1.5);
        h3=plot(1:100,RWinh,'m','linewidth',1.5);       
        lgd=legend([h1,h2,h3],{'R','Wexc','R*Winh'});
    else
        lgd=legend(h1,'R');
    end
    lgd.ItemTokenSize = [5,1];
    ylimit=ylim; ylim([-1 ylimit(2)]);

    subplot(3,4,11)
    plot(reshape(spike_times(:,:,sel)+ref.model.axonalDelay,100*ref.model.maxES,[]),repmat(1:100,[1,ref.model.maxES]),'.');hold on;
    ylim([-0.5 100]);xlim([15 50])
    ylabel('Motor unit');
    xlabel('Time (ms)')
    title('MU trigger time')
    legend(legend_labels,'location','northeast');%legend(num2str(ref.intensities(sel)'),'location','northeast')
    
    subplot(3,4,12);hold on;box on;
    for i=sel
       [N,edges] = histcounts(spike_times(:,:,i)+ref.model.axonalDelay,15:2:50,'Normalization','countdensity');
       edges = edges(2:end) - (edges(2)-edges(1))/2;
       plot(edges, N,'linewidth',1.5);
    end
    xlim([15 50]);
    ylimit=ylim; ylim([-1 ylimit(2)]);
    lgd=legend(legend_labels,'location','northeast');%legend(num2str(ref.intensities(sel)'),'location','northeast')
    lgd.ItemTokenSize = [5,1];
    title('Histogram');xlabel('Trigger time (ms)');ylabel('Count density')
    
    set(gcf,'position',[50 50 800 600])
    
end

File: plot_param_panel.py
import numpy as np
import matplotlib.pyplot as plt

def plot_param_panel(p, ref):
    # Mapping variables from the ref dictionary using tuple keys for nested structures
    spike_times  = ref['sim']['spike_times']
    NRMSD        = ref['NRMSD']
    R            = ref['model']['R']
    Wexc         = ref['model']['Wexc']
    RWinh        = ref['model']['RWinh']  # R*Winh
    simMEP       = ref['sim']['simMEP2']
    DIwave       = ref['model']['DIwave']
    t            = ref['model']['t']
    y0           = ref['y0']
    intensities  = ref['intensities']
    subj         = ref['subj']
    tcrop        = ref['tcrop']
    
    # Optional/conditional fields
    boundary     = ref['model']['boundary'] if 'boundary' in ref['model'] else None
    boundarytext = ref['model']['boundarytext'] if 'boundarytext' in ref['model'] else []
    axonalDelay  = ref['model']['axonalDelay'] if 'axonalDelay' in ref['model'] else 0
    maxES        = ref['model']['maxES'] if 'maxES' in ref['model'] else 1
    withRC       = ref['model']['withRC'] if 'withRC' in ref['model'] else False

    nParams = len(p)
    plt.figure(figsize=(10, 8)) # Replicating set(gcf,'position',[50 50 800 600]) ratio

    # --- Top Row: Parameter value indicators ---
    for i in range(nParams):
        plt.subplot(3, nParams, i + 1)
        rescale = 0.9
        if boundary is not None:
            # Map p into 0~1 range for visualization
            locP = rescale * (p[i] - boundary[i, 0]) / (boundary[i, 1] - boundary[i, 0])
        else:
            locP = 0.5 * rescale
        
        plt.scatter([1], [locP], color='C0', zorder=5) # "filled" scatter
        plt.text(1.01, locP-0.05, f"{p[i]:.2f}")
        
        if boundary is not None:
            plt.scatter([1, 1], [0, rescale], marker='_', color='black')
            plt.plot([1, 1], [0, rescale], 'k')
            plt.text(1, 0-0.03, str(boundary[i, 0]), ha='center', va='top')
            plt.text(1, rescale+0.03, str(boundary[i, 1]), ha='center', va='bottom')
            
        plt.ylim([-0.1, 1.1])
        plt.axis('off')
        if boundarytext:
            plt.title(boundarytext[min(i, len(boundarytext)-1)], fontsize=9)

    # --- Middle Row: MEP comparison across intensities (Subplot 312) ---
    plt.subplot(3, 1, 2)
    # Using 'F' order (column-major) to flatten arrays to match MATLAB's (:) operator
    y0_flat = y0.flatten(order='F')
    simMEP_flat = simMEP.flatten(order='F')
    
    plt.plot(y0_flat, 'k', linewidth=2, label='MEP')
    plt.plot(simMEP_flat, 'r', linewidth=1, label='simMEP')
    
    # R^2 = 1 - sumsqr(err) / sumsqr(y - mean(y))
    R2 = 1 - (np.sum((y0_flat - simMEP_flat)**2) / np.sum((y0_flat - np.mean(y0_flat))**2))
    print(f'R^2 = {R2}')
    plt.title(f"MEPs of subject {subj} \n(R^2 = {R2:.2g}, NRMSD = {NRMSD*100:.2g}%)")

    # Draw vertical lines for intensity boundaries
    # size(simMEP, 1) in MATLAB refers to the number of points per intensity block
    pts_per_intensity = simMEP.shape[0]
    for i in range(simMEP.shape[1]):
        plt.axvline(x=i * pts_per_intensity, color='black', linewidth=0.5)
    
    plt.xlim([0, len(simMEP_flat)])
    plt.xlabel(f"Time ({tcrop[0]}-{tcrop[1]} ms)")
    plt.xticks([]) # Equivalent to xticklabel, []
    
    ylimit = plt.gca().get_ylim()
    for i in range(len(intensities)):
        plt.text(pts_per_intensity * (i + 0.5), ylimit[1] * 0.5, 
                 f"{int(intensities[i])}%\nMSO", ha='center', fontsize=8)
    
    plt.legend(loc='center left', prop={'size': 8}) # 'west' equivalent
    plt.ylabel('Amplitude (mV)')

    # Select 3 intensities to show in detail [Last, Middle, First]
    sel = [len(intensities) - 1, int(round(len(intensities) / 2)) - 1, 0]
    legend_labels = [f"{int(intensities[s])}%" for s in sel]

    # --- Subplot 9: DI-waves ---
    plt.subplot(3, 4, 9)
    plt.plot(t, DIwave[sel, :].T, linewidth=1.5)
    plt.xlabel('Time (ms)')
    plt.title('DIwaves')
    plt.xlim([0, 20])
    curr_yl = plt.gca().get_ylim()
    plt.ylim([-0.05, curr_yl[1]])
    lgd = plt.legend(legend_labels, loc='upper right', fontsize=7)
    lgd.get_frame().set_linewidth(0.5)

    # --- Subplot 10: Model Parameters across Neurons ---
    ax1 = plt.subplot(3, 4, 10)
    neurons = np.arange(1, 101)
    h1, = plt.plot(neurons, R, 'k.', label='R')
    plt.title('Model param.')
    plt.xlabel('Motor neuron')
    # Highlight specific neurons [1, 10, 20, 60, 100]
    hi_idx = [0, 9, 19, 59, 99]
    plt.plot(neurons[hi_idx], R[hi_idx], 'ko', fillstyle='none')
    
    if withRC:
        h2, = plt.plot(neurons, Wexc, 'b', linewidth=1.5, label='Wexc')
        h3, = plt.plot(neurons, RWinh, 'm', linewidth=1.5, label='R*Winh')
        plt.legend(handles=[h1, h2, h3], fontsize=7)
    else:
        plt.legend(handles=[h1], labels=['R'], fontsize=7)
    ax1.set_xlim([0, 100])
    plt.ylim([-1, plt.gca().get_ylim()[1]])

    # --- Subplot 11: MU Trigger Time (Raster Plot) ---
    plt.subplot(3, 4, 11)
    # spike_times is (100, maxES, nIntensities)
    for s_idx in sel:
        # Replicate MATLAB's reshape/repmat logic for scatter plotting
        st = (spike_times[:, :, s_idx] + axonalDelay).flatten()
        units = np.repeat(np.arange(1, 101), maxES)
        mask = ~np.isnan(st)
        plt.plot(st[mask], units[mask], '.', markersize=2)
        
    plt.ylim([-0.5, 100])
    plt.xlim([15, 50])
    plt.ylabel('Motor unit')
    plt.xlabel('Time (ms)')
    plt.title('MU trigger time')
    plt.legend(legend_labels, loc='upper right', fontsize=7)

    # --- Subplot 12: Trigger Time Histogram ---
    plt.subplot(3, 4, 12)
    bins = np.arange(15, 51, 2)
    for s_idx in sel:
        data = (spike_times[:, :, s_idx] + axonalDelay).flatten()
        data = data[~np.isnan(data)]
        # count density = count / bin_width
        counts, edges = np.histogram(data, bins=bins)
        centers = edges[:-1] + np.diff(edges) / 2
        plt.plot(centers, counts / np.diff(edges), linewidth=1.5)
        
        
    plt.xlim([15, 50])
    plt.ylim([-1, plt.gca().get_ylim()[1]])
    plt.title('Histogram')
    plt.xlabel('Trigger time (ms)')
    plt.ylabel('Count density')
    plt.legend(legend_labels, loc='upper right', fontsize=7)

    plt.subplots_adjust(hspace=0.4,wspace=0.5)
    plt.show()